### PR TITLE
feat: Helm chart to deploy unpoller in k8s

### PR DESCRIPTION
**Pull request by jlpedrosa**
Initial version of a helm chart to deploy unpoller in kubernetes.

* Deployment for unpoller
* Secret for the unpoller configuration
* PodMonitor for unpoller deployment (Depends on PrometheusOperator)
* Optional Dashboards that import dashboards from https://grafana.com/grafana/dashboards/ in prometheus mode
* Optional Grafana Instance to persist the Dashboards
* Makefile target that generates Helm documentation.

One this PR is ready, two more PRs needs to be created:
* CI to publish the dashboards (there are multiple ways to skin the cat, and depends on where the maintainers of the repo want to keep the artifacts, GH branch (usual approach for small projects) GH packages, external OCI storage).
* Update the documentation in how to deploy Unpoller in k8s).

cc: @platinummonkey

### CODE REVIEW

**Comment by davidnewhall** [comment]
I feel like all of this should be in another repo.

File: charts/unpoller/Chart.yaml
Comment (davidnewhall) [inline charts/unpoller/Chart.yaml:9]: What are these?

File: charts/unpoller/Chart.yaml
Comment (davidnewhall) [inline charts/unpoller/Chart.yaml:16]: This is part of why this feels out of place. I don't think the repo maintainers can also maintain this with high confidence. Whatcha think Cody?

File: charts/unpoller/README.md
Comment (davidnewhall) [inline charts/unpoller/README.md:3]: This feels like a quickly-outdated place to put a release version.

File: charts/unpoller/README.md
Comment (davidnewhall) [inline charts/unpoller/README.md:7]: unpoller.com*

File: charts/unpoller/README.md
Comment (davidnewhall) [inline charts/unpoller/README.md:13]: Right?

File: charts/unpoller/templates/dashboards.yaml
Comment (davidnewhall) [inline charts/unpoller/templates/dashboards.yaml:12]: Probably don't want the revision in these urls.

**Review by davidnewhall — CHANGES_REQUESTED** [review]
There's a few piece of data missing here. I mentioned them with inline comments. My thought is that we should make a new repo in the unpoller order. Maybe `helm-chart`? Is this actually a helm chart? I'm not a k8s users so this is mostly foreign to me.

If we create a new repo, we can give editor access to the actual maintainer(s) of this new "code."

File: charts/unpoller/README.md
Comment (jlpedrosa) [reply charts/unpoller/README.md:3]: versioning should be updated automatically in CI (Seperate PR), I created a make target that generates this file (Including the versios), the idea is trigger it during the release process, I haven't yet had a look at your CI to see how it can be integrated.

File: charts/unpoller/README.md
Comment (jlpedrosa) [reply charts/unpoller/README.md:13]: Yes, it is autogenerated with whatever is on Chart.yaml, once we decide what goes in there it will be re-generated.

File: charts/unpoller/templates/dashboards.yaml
Comment (jlpedrosa) [reply charts/unpoller/templates/dashboards.yaml:12]: I agree, but I could not find a URL to download it that did not include the revision:
`https://grafana.com/api/dashboards/11314/download` => 404

If you have maybe the source of the repos I can point to them?

File: charts/unpoller/Chart.yaml
Comment (jlpedrosa) [reply charts/unpoller/Chart.yaml]: I deny having copy pasted the structure from other chart and forgetting fixing it. :rofl: 
fixed.

**Comment by jlpedrosa** [comment]
> There's a few piece of data missing here. I mentioned them with inline comments. My thought is that we should make a new repo in the unpoller org. Maybe `helm-chart`? Is this actually a helm chart? I'm not a k8s users so this is mostly foreign to me.
> 
> If we create a new repo, we can give editor (write) access to the actual maintainer(s) of this new "code."

Hi! so this is completely up to the maintainers, from my perspective makes sense to collocate code & deployment, same as the docker for the container is together. Said that there are some projects that split it into different repos, what ever you guys prefer is cool by me and I can create the PR in any repo. I think collocating it gives a sense of "official" and avoids more and more people doing [their charts](https://artifacthub.io/packages/search?ts_query_web=unpoller&sort=relevance&page=1).

From the maintenance perspective, kube api resources used here (except for grafana) are pretty stable, I am happy to help if you tag me on PRs related to this.

File: charts/unpoller/README.md
Comment (jlpedrosa) [reply charts/unpoller/README.md]: fixed

File: charts/unpoller/templates/dashboards.yaml
Comment (davidnewhall) [reply charts/unpoller/templates/dashboards.yaml:12]: `https://grafana.com/api/dashboards/11314/revisions/latest/download`

**Comment by davidnewhall** [comment]
Here's my hangup... You've added like 7 or 8 yaml files, a go template and a README that doesn't say anything. I have no idea what any of these new files accomplish. No idea what they do. Not sure who uses them, when or why. How can I review this contribution and tell you if these new files work or not? Nothing tells me what I'm supposed to do with them, how I test them, or  if they're even included on purpose.

You mention that deployments should live next to the code. I generally agree. I do not see a deployment in this contribution, so I'm still wondering if these new files make sense in their own repo. If you provide more information about the end game this will probably make more sense to me.

File: charts/unpoller/templates/dashboards.yaml
Comment (jlpedrosa) [reply charts/unpoller/templates/dashboards.yaml:12]: :facepalm:  Thanks, I'll update it.

**Comment by jlpedrosa** [comment]
> Here's my hangup... You've added like 7 or 8 yaml files, a go template and a README that doesn't say anything. I have no idea what any of these new files accomplish. No idea what they do. Not sure who uses them, when or why. How can I review this contribution and tell you if these new files work or not? Nothing tells me what I'm supposed to do with them, how I test them, or if they're even included on purpose.
> 
> You mention that deployments should live next to the code. I generally agree. I do not see a deployment in this contribution, so I'm still wondering if these new files make sense in their own repo. If you provide more information about the end game this will probably make more sense to me.

hey @davidnewhall! Yeah! totally get you, absolutely fair. All this sutff if very blurry if you are not super familiar with it!
Typically most of the helm documentation is based on the parameters that can be configured in the chart as the core concepts of Deployment/Service/Ingress etc are usually "table stakes" for the kubernetes people, at least in simple deployments as this. Said that it won't hurt add more documentation to the Readme and it's always better. let me do that. Helm is essentially a templating tooling. Gets the yamls in templates and applies substitutions from `values.yaml` and the metadata is `Chart.yaml`. More precisely go templating. So TL;DR; Helm is used by people/teams that want to deploy unpoller in kubernetes, it takes all the rendered templates and pushes them to kubernetes api. Each object Service/Deployment/... is a known object for kube-api.

About the endgame, the helm "deployment" process, consist of two steps 1) add a "repository" with the additional helm charts (unpoller in this case) `helm repo add`  and  2) proceed with the installation (helm install).

Repository: The two simplest way to have a helm repository that is one of the two ways:
`* Option 1`: [GH Pages](https://docs.github.com/en/pages/getting-started-with-github-pages/creating-a-github-pages-site) + helm-chart-releaser action: This configures GH with an specific branch on your git repository (usually gh-pages) with the files generated in the CI which is an index.yaml and the tar.gz with the chart. If you want an example of how the repo looks like [here](https://github.com/jlpedrosa/zigbee2mqtt-chart/settings/pages) it is.

`* Option 2: GH Packages`:  as the charts are a tar.gz is possible to use the GH docker repository as an oci repository to store the helm charts for example:  oci://ghcr.io/grafana/helm-charts, so the CI would push the file to a new package (unpoller-chart name as unpoller is in use) in this packages repo. Funny enough in this route, there's no `index.yaml` and requires helm 3.8 (or later) to install using the chart directly.

So the endgame would be:
* Decision: Keep the helm chart in this repo or create a new repo and repeat this PR on the new repo `unpoller-chart` or what ever name you decide.
* Create the CI Process with` Option 1` or `Option 2` (for "deployments" options in the paragraph above),  after you make the decision (I can code both) in a new PR
* Create the PR with the documentation and instructions for unpoller.com (but don't merge it)
* Test the full release cycle, ensure everything is working fine (I can do that)
* Merge the documentation PR.

PS: Thank you for your project and time to review this, I am doing the "same" PR in few upstreams and I clearly was too light on the descriptions/docs for the PR.  To give you more context, this PR is inspired by this thread https://github.com/unpoller/unpoller/issues/410

File: charts/unpoller/README.md
Comment (jlpedrosa) [reply charts/unpoller/README.md:3]: added a bit more docs, let me know if that is enough for this section, I was planning to do a separated PR for the unpoller.com site

File: charts/unpoller/templates/dashboards.yaml
Comment (jlpedrosa) [reply charts/unpoller/templates/dashboards.yaml]: fixed.

File: charts/unpoller/values.yaml
Comment (platinummonkey) [inline charts/unpoller/values.yaml:9]: might actually make more sense to just leave it as `"v2"` this should update with the semantic versioning

File: charts/unpoller/Chart.yaml
Comment (platinummonkey) [inline charts/unpoller/Chart.yaml:15]: This chart version we could also version separately just as `"1.0.0"`

**Comment by platinummonkey** [comment]
> If we create a new repo, we can give editor (write) access to the actual maintainer(s) of this new "code."

A new repo might be helpful, will create that one now

**Comment by platinummonkey** [comment]
moving over here https://github.com/unpoller/helm-chart

File: charts/unpoller/Chart.yaml
Comment (davidnewhall) [reply charts/unpoller/Chart.yaml]: Very happy to hear that! I'll mostly bow out of this then and only poke my nose in when I'm curious.

File: charts/unpoller/README.md
Comment (davidnewhall) [reply charts/unpoller/README.md:3]: Thank you for the additional context and documentation. This all seems fine with me. I trust @platinummonkey will help you guide this to a high bar contribution. Good luck!

**Pull request closed without merging**
